designs/resource_metadata_overriding.md

```diff
@@ -0,0 +1,170 @@
+the template and for each resource reading the Metadata and replacing values as specified.
+
+In the future, we can support creating these templates from the different frameworks in a command directly within 
+SAM CLI but is out of scope in the initial implementation of support.
+
+Success criteria for the change
+-------------------------------
+
+* Ability to invoke functions locally that contain Metadata on a Resource
+* Process a template with CDK Metadata on a resource.
+
+Out-of-Scope
+------------
+
+* A command that will generate the template from the framework.
+
+User Experience Walkthrough
+---------------------------
+
+CDK is a framework that appends this Metadata to Resources within a template and will use this as an example.
+
+### Customer using CDK
+
+A customer will use CDK to generate the template. This can be done by generating a template and saving it to a file:
+`cdk synth > template.yaml`. Then will then be able to `sam local [invoke|start-api|start-lambda]` any 
+function they have defined [1].   
+
+
+[1] Note: The cdk version must be greater than v0.21.0 as the metadata needed to parse is not appended on older versions. 
+
+
+Implementation
+==============
+
+CLI Changes
+-----------
+
+For the features currently in scope, there are no changes to the CLI interface.
+
+### Breaking Change
+
+No breaking changes
+
+Design
+------
+
+All the providers, which are used to get resources out of the template provided to the command, call 
+`SamBaseProvider.get_template(template_dict, parameter_overrides)` to get a normalized template. This function call is 
+responsible for taking a SAM template dictionary and returning a cleaned copy of the template where SAM plugins have 
+been run and parameter values have been substituted. Given the current scope of this call, expanding it to also normalize 
+metadata, seems reasonable. We will expand `SamBaseProvider.get_tempalte()` to call a `TemplateMetadataNormalizer` class
+that will be responsible for understanding the metadata and normalizing the template with respect to the metadata. 
+
+Template snippet that contains the metadata SAM CLI will parse and understand.
+
+```yaml
+
+Resources:
+    MyFunction:
+      Type: AWS::Lambda::Function
+      Properties:
+        Code: 
+          S3Bucket: mybucket
+          S3Key: myKey
+        ...
+      Metadata:
+        aws:asset:path: '/path/to/function/code'
+        aws:asset:property: 'Code'
+```
+
+The two keys we will recognize are `aws:asset:path` and `aws:asset:property`. `aws:asset:path`'s value will be the path
+to the code, files, etc that are help on the machine, while `aws:asset:property` is the Property of the Resource that
+needs to be replaced. So in the example above, the `Code` Property will be replaced with `/path/to/function/code`.
+
+Below algorithm to do this Metadata Normalization on the template.
+
+```python
+class TemplateMetadataNormalizer(object):
```

Comment:
Is this normalizer going to be injected? can we have more than one normalizer such that we can chain them?

Reply:
This will not be injected. The Method is static. I wanted to keep this out of the `SamBaseProvider` class, so it can make the decision on when/where to run this in the flow of getting the `template_dict`. It is possible to have more than one normalizer but they could happen at different times. For now, I elected to keep it simple and just make the method call to this class where we needed.